rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -268,6 +271,221 @@ impl CanisterHistory {
     }
 }
 
+/// `TaskQueue` represents the implementation of queue structure for tasks satisfying  conditions:
+///
+/// 1. If there is a `Paused` or `Aborted` task it will be returned first.
+/// 2. If an `OnLowWasmMemoryHook` is ready to be executed, it will be returned next.
```

Comment:
I don't see that happening anywhere. As in, `pop()` never returns `ExecutionTask::OnLowWasmMemory`. OTOH, we still have a `get_on_low_wasm_memory_hook_status()` method, which IMHO should not be part of the `TaskQueue` API.

Reply:
You are right, the logic from this comment is implemented in the follow-up, I updated the comment.

Reply:
Regarding the `get_on_low_wasm_memory_hook_status()` as mentioned in the comment above, it is used only for the checkpointing to extract the private value from the struct. What alternative do you propose assuming that we keep the field private?

Reply:
I renamed `get` to `peek` method.

Reply:
The alternative I propose is to define a `TaskQueue` proto and define the conversion to / from proto here instead of exposing private fields piecemeal just so they can be encoded / decoded by `CanisterStateBits`.

It's what we do with virtually all parts of `ReplicatedState` apart from the monster that `CanisterStateBits` has grown into.

If you want a struct whose state is entirely private and its logic self-contained, then you need to make sure that the invariants hold when you load it; and that clients cannot use methods that are "only supposed to be used by the persistence logic" to arbitrarily mess with its internals. Otherwise, it's more honest to have the fields be public so clients are aware of what they are dealing with.

Reply:
We can create a `CanisterStateBits` version of this struct can be done after this PR. We already have implemented conversion to `CanisterStateBits`, and the point of this PR is to refactor `TaskQueue`, so I guess that should not be the blocker for this PR.

We can come back to this and decide if it makes sense to do so, but it will require the process of 3 releases in order to move this. 

Does that makes sense to you?

Reply:
Sure. You can do it in 2 steps, though:
1. Define the new `TaskQueue` message (containing the queue plus OnLowWasmMemory task status; later also the separate paused/aborted task field). Populate both the `TaskQueue` message and the existing fields when encoding. Read from either the new message or the old fields when loading.
2. Drop the old fields. Stop encoding them. Stop reading from them.

Reply:
I agree that this PR should not be blocked on how things are encoded to proto. The fields were separate, so they were encoded separately and now the PR is trying to refactor to bring them together. We should move into the state where they are encoded as part of the single struct but that doesn't need to happen here.

Reply:
Thank you. The follow-up is created https://dfinity.atlassian.net/browse/EXC-1752.